stablehlo/reference/Element.cpp

```diff
@@ -173,6 +173,221 @@ bool areApproximatelyEqual(APFloat f, APFloat g) {
+    // Combines 'resultBitWidth' / 'operandBitWidth' elements into one element,
+    // appending the real and imaginary parts in little endian representation.
+    // TODO(#1460): Explore adding support for big endian architectures
+    auto applyComplexConversion = [&](auto offset) {
+      auto binary = APInt(resultBitWidth, 0);
+      for (auto i = 0; i < resultBitWidth / (2 * operandBitWidth);
+           ++i, ++offset) {
+        binary = leftShiftAndMask(
+            binary, operandBitWidth,
+            elements[offset].getComplexValue().imag().bitcastToAPInt().zext(
+                resultBitWidth));
+        binary = leftShiftAndMask(
+            binary, operandBitWidth,
+            elements[offset].getComplexValue().real().bitcastToAPInt().zext(
+                resultBitWidth));
+      }
+      return binary;
+    };
+    auto baseType = type.cast<ComplexType>().getElementType().cast<FloatType>();
+    auto binaryReal = applyComplexConversion(0);
+    auto binaryImag =
+        applyComplexConversion(resultBitWidth / (2 * operandBitWidth));
+    return Element(type,
+                   std::complex<APFloat>(
+                       APFloat(baseType.getFloatSemantics(), binaryReal),
+                       APFloat(baseType.getFloatSemantics(), binaryImag)));
+  }
+
+  report_fatal_error(invalidArgument("Unsupported conversion from %s to %s",
+                                     debugString(operandType).c_str(),
+                                     debugString(type).c_str()));
+}
+
+// Returns bitcast converted larger bitwidth Element object to array of smaller
+// bitwidth Element objects.
+SmallVector<Element> bitcastConvertOneToMany(Type type, const Element &el) {
+  SmallVector<Element> results;
+  auto operandType = el.getType();
+  int64_t operandBitWidth = isSupportedComplexType(operandType)
+                                ? operandType.cast<ComplexType>()
+                                      .getElementType()
+                                      .cast<FloatType>()
+                                      .getIntOrFloatBitWidth()
+                                : operandType.getIntOrFloatBitWidth();
+  int64_t resultBitWidth = isSupportedComplexType(type)
+                               ? type.cast<ComplexType>()
+                                     .getElementType()
+                                     .cast<FloatType>()
+                                     .getIntOrFloatBitWidth()
+                               : type.getIntOrFloatBitWidth();
+
+  // Splits element into 'operandBitWidth' / 'resultBitWidth' elements, converts
+  // to Element object of type 'type', and appends them to 'results'.
+  auto applyIntOrFloatConversion = [&](APInt mask) {
+    for (auto i = 0; i < operandBitWidth; i += resultBitWidth) {
+      auto binary = leftShiftAndMask(
+          APInt(resultBitWidth, 0), resultBitWidth,
+          mask.extractBits(resultBitWidth, i).zext(resultBitWidth));
```

Comment:
extractBits generates a bitstring os size `resultBitWidth`. Zero-extending it by `resultBitWidth` seems like a no-op.